zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbFormState.java

```diff
@@ -212,6 +213,40 @@ public void deleteFormInFormKeyByFormIdAndVersionTagColumnFamily(final FormRecor
     formKeyByFormIdAndVersionTagColumnFamily.deleteIfExists(tenantAwareFormIdAndVersionTagKey);
   }
 
+  @Override
+  public void setMissingDeploymentKey(
+      final String tenantId, final long formKey, final long deploymentKey) {
+    tenantIdKey.wrapString(tenantId);
+    dbFormKey.wrapLong(formKey);
+    dbDeploymentKey.wrapLong(deploymentKey);
+
+    final var form = formsByKey.get(tenantAwareFormKey);
+    if (form.getDeploymentKey() == deploymentKey) {
+      return;
+    }
+
+    if (form.hasDeploymentKey()) {
+      throw new IllegalStateException(
+          String.format(
+              "Expected to set deployment key '%d' on form with key '%d', but form already has deployment key '%d'.",
+              deploymentKey, formKey, form.getDeploymentKey()));
+    }
+    formVersion.wrapLong(form.getVersion());
+
+    form.setDeploymentKey(deploymentKey);
+    formsByKey.update(tenantAwareFormKey, form);
+    formByIdAndVersionColumnFamily.update(tenantAwareIdAndVersionKey, form);
+    formKeyByFormIdAndDeploymentKeyColumnFamily.insert(
+        tenantAwareFormIdAndDeploymentKey, fkFormKey);
+
+    final var dbKey = new TenantIdAndFormId(tenantId, BufferUtil.bufferAsString(form.getFormId()));
+    final var formInCache = formsByTenantIdAndIdCache.getIfPresent(dbKey);
+    if (formInCache == null || formInCache.getVersion() <= form.getVersion()) {
+      final var copyForCache = form.copy();
+      formsByTenantIdAndIdCache.put(dbKey, copyForCache);
+    }
```

Comment:
```suggestion
    final var formId = BufferUtil.bufferAsString(form.getFormId());
    getFormFromCache(tenantId, formId)
        .filter(cachedForm -> cachedForm.getVersion() == form.getVersion())
        .ifPresent(
            cachedForm -> formsByTenantIdAndIdCache.put(new TenantIdAndFormId(tenantId, formId), form));
```

🔧 Just a small suggestion. What's more important is that the check you have now could cause issues. The way it works if we only have the latest version of a form cached, which I have opinions on but we shouldn't tackle those in this PR.

If this reconstructor intermittently adds forms to the cache which may or may not be the latest, it could be that a form retrieved in between finds a non-latest form in the cache which results in unexpected behavior.
So the check should be:
- Is there a form in the cache?
    - Yes, does this form have the latest version?
        - Yes, update the form in the cache
        - No, don't do anything
    - No, don't do anything

Reply:
Oh you are right! I assumed the last version would be there, but it's a cache nonetheless :-D